libnmstate/nm/device.py

```diff
@@ -277,6 +277,19 @@ def _delete_device_callback(src_object, result, user_data):
                 e,
             )
             mainloop.execute_next_action()
+        elif (
+            e.matches(
+                nmclient.NM.DeviceError.quark(),
+                nmclient.NM.DeviceError.NOTSOFTWARE,
+            )
+            and nmdev.is_software()
+        ):
+            logging.debug(
+                'Device %s is software ignoring race condition: error=%s',
```

Comment:
What exactly is the race here? Can the interface already be removed by NM before the command reaches NM?

Reply:
This is probably the case for interfaces that were created by NM in the first place.

Reply:
So if this is the race, please explain it in the commit message and the error message. Also to be sure, there should be a check if the device is already deleted instead of assuming that any error message means that it is deleted.

Reply:
The handling of the race could be tested by trying to remove the same interface twice.

Reply:
I don't know where to find out if the interface was already removed. Can you point me the right direction?

Reply:
actually I just saw that a few lines above your code there is a check that is supposed to check if a device was already deleted (it does not look like it explicitly checks this, though).

@thom311  can you tell us how to properly check in the callback of a `NM.Device.delete_async()` call whether the device is deleted in case the call fails? I see that there is a a `is_real()` check: https://lazka.github.io/pgi-docs/#NM-1.0/classes/Device.html#NM.Device.is_real - can it be called in this case?

Reply:
> whether the device is deleted in case the call fails?

if the call fails, you don't know whether the deletion succeeded. That depends on why it failed...

Overall, it should be that

 - if a device exists and is-real, then it should be tracked by `NMClient.get_devices()` property.
 - if the device exists and is either real or not real, it should be tracked by `NMClient.get_all_devices()`.
 - after `delete_async()` completes with success, the device should no longer be real and be deleted from `NMClient.get_devices()`.

However, libnm itself does not enforce such consistency. In many aspects, it exposes what is on D-Bus, so theoretically `NMClient.get_devices()` could also track devices that are not real (though, of course, that probably would indicate a bug). You don't have to assume that the consistency is violated, because you have to trust NMClient (and NetworkManager) to not expose you garbage information (otherwise, what would you do anyway to know the truth?).

Note that NMClient also has signals "any-device-added", "any-device-removed" (for `NMClient.get_all_devices()`, and "device-added", "device-removed" (for `NMClient.get_devices()`). And of course, there are also GObject properties "devices" and "all-devices", so you could also subscribe to "notify::devices" and "notify::all-devices" signals.

In general, I think you should not care about unrealized devices (!is_real). Just ignore `NMClient.get_all_devices()`.

The best way to know when a device is gone (or unrealized) is

- listen to `NMClient` "device-removed" signal.
- look for the device in `NMClient.get_devices()`.

Reply:
> > whether the device is deleted in case the call fails?
> 
> if the call fails, you don't know whether the deletion succeeded. That depends on why it failed...

[...]

> The best way to know when a device is gone (or unrealized) is
> 
> * listen to `NMClient` "device-removed" signal.
> * look for the device in `NMClient.get_devices()`.

The problem is, that the device might be deleted by NM automatically instead of as a result of the async delete call. So if the call fails but the device is deleted, then 
the error can be ignored. Does NM tell that the call failed because it does not exist? If this is the case, Nmstate could just check for this. If not, then from the options you outlined, it seems that checking if the device is in `NMClient.get_devices()` is the way to go. Listening for signals at that point does not work because the deletion would already have happened. Also if the device is not deleted, this would take forever.

What I do not understand is, why `is_real()' cannot be used. Since it is the callback for a method of a device, we already have the device object, so `nmdevice in NMClient.get_devices()` should return the same result as `nmdevice.is_real()`. What am I missing?

Reply:
> Listening for signals at that point does not work because the deletion would already have happened.

It works. You already have a `NMDevice` instance that you know is-real, and that you are about to call `device.delete_async()` on. If you register to `NMClient.device-removed` signal, and you encounter a delete signal for your device, you know it's gone. Of course, you can also check with `any([device is d for d in nmc.get_devices ()])` (which is a bit more expensive, but a failure of `delete_async()` is not the common case).

Theoretically, the error reason will also tell you that the device no longer existed, but that isn't useful here. You are not really interested in success of the `delete_async()` call. You are interested to get rid of the device. You first do something that hopefully will rid it (`delete_async()`) and afterwards you care whether it's still there or not. So, if the call fails with some other failure, but the device is actually gone, what did you gain by caring about the result?

> What I do not understand is, why `is_real()` cannot be used.

is_real() is a property of the NMDevice instance. When an instance gets removed from the cache (`NMClient.get_devices()`), it's not clear that all properties should be cleared. Would it be useful that the device's `ifname` gets set to NULL the moment the instance gets removed from the cache? I don't think so. Maybe we should add a property `nm_object_is_alive()`, which returns TRUE as long as the object is still in the cache. Then you could actually check `not device.is_alive() or not device.is_real()`...

Reply:
I have to admit that I totally started to loosing track of what is going on.